### PR TITLE
Run the float head of QAT TensorRT engines in FP16

### LINKED ISSUE #26159: QAT INT8 engine ~21% slower than PTQ INT8 (and even PTQ FP16) at imgsz=416
### Question

Following the QAT recipe merged in #26083, I fine-tuned a base checkpoint with `quantize=8` and exported it to a TensorRT INT8 engine (**Engine A**), and compared it against **Engine B** (FP32), **Engine C** (PTQ-FP16), and **Engine D** (PTQ-INT8), all exported from the same base checkpoint. mAP came out as expected (QAT-INT8 ≈ PTQ-FP16, far better than PTQ-INT8). However, **FPS is the worst among all precisions except FP32** — even slower than PTQ-FP16.

**mAP50-95 (`yolo26m`, imgsz=416):**

| Engine | mAP |
| --- | --- |
| FP32 | 0.783 |
| PTQ-FP16 (currently in use) | 0.779 |
| PTQ-INT8 | 0.731 |
| QAT-INT8 | 0.779 |

**FPS (same engines):**

| Engine | FPS (speedup vs FP32) |
| --- | --- |
| FP32 | 536.3 (1.00x) |
| PTQ-FP16 | 1070.5 (2.00x) |
| PTQ-INT8 | 1295.5 (2.42x) |
| QAT-INT8 | 845.7 (1.58x) |

So QAT-INT8, despite being INT8, is **about 21% slower than the PTQ-FP16 engine we currently use**, and falls far short of PTQ-INT8's speedup.

**My hypothesis:** Per #26083, QAT keeps the output head in float to preserve accuracy ("Leave the head in float during QAT") — only the backbone is quantized. I suspect this mixed-precision boundary (INT8 backbone → FP32 head) forces TensorRT to insert additional reformat/cast layers at that boundary, and this transition overhead offsets the speed gain from the INT8 backbone. PTQ-INT8, by contrast, appears to be quantized more uniformly (at the cost of the accuracy loss shown above).

**Questions:**

1. Is this an intended trade-off of the current QAT implementation (a design choice to keep the head in float), or is this a regression or a suboptimally built engine? This seems unusual, since QAT is generally understood as a technique to gain speed while minimizing accuracy loss.
2. Could you recommend a standard diagnostic method to confirm exactly where this reformat overhead is occurring (e.g., per-layer profiling via `trtexec --verbose`)?
3. Is quantizing the head as well (to INT8 or FP16) planned or possible with the current `quantize=8` QAT path? If so, what accuracy/speed trade-off would you expect?
4. PTQ-INT8 gains only a 0.42x speedup over PTQ-FP16 while losing 4.8 mAP, and QAT-INT8 matches PTQ-FP16's mAP but is actually slower. Is there any configuration where QAT-INT8 would actually beat PTQ-FP16 on FPS while maintaining mAP parity? Or, until head quantization is supported, is sticking with PTQ-FP16 the right call for this model/imgsz combination for now?

### Additional Information

- Model: `yolo26m`, imgsz=416
- Export method
    - Export code for FP32, PTQ-FP16, and QAT-INT8

```python
    # export.py
    from ultralytics import YOLO

    model = YOLO("runs/train/d12345_s1_imgsz416/best.pt")
    model.export(
        format="engine",
        quantize=32,                    
        imgsz=416,
        batch=1,          
        device=0,
    )
```

    - PTQ-INT8 export code

```python
    from ultralytics import YOLO

    model = YOLO("runs/train/d12345_s1_imgsz416/weights/best.pt")
    model.export(
        format="engine",
        quantize=8,                     
        data="/data/Projects/ultralytics/data/dataset/calib.yaml",
        imgsz=416,
        batch=1,                        
        device=0,
    )
```

- mAP measurement method

```python
    # tested by swapping the engine path
    from ultralytics import YOLO

    model = YOLO("./runs/train/d12345_s1_imgsz416/PTQ-FP16/best.engine")
    metrics = model.val(
        data="/data/Projects/ultralytics/data/dataset/dataset_v1_v2_v3_v4_v5_yolo/data.yaml",
        split="test",
        imgsz=416,
        batch=1,
        name="ptq_fp16",
    )
    print(metrics.box.map, metrics.box.map50)
```

- FPS measurement method

```python
    # bench_fixed.py
    import json, struct, time
    from pathlib import Path

    import numpy as np
    import tensorrt as trt
    import torch

    IMGSZ, WARMUP, ITERS = 416, 500, 2000
    ENGINES = {
        "PTQ-FP32": "runs/train/d12345_s1_imgsz416/PTQ-FP32/best.engine",
        "PTQ-FP16": "runs/train/d12345_s1_imgsz416/PTQ-FP16/best.engine",
        "PTQ-INT8": "runs/train/d12345_s1_imgsz416/PTQ-INT8/best.engine",
        "QAT-INT8-head-fp32": "runs/train/d12345_s1_imgsz416_qat/weights/QAT-int8-head-fp16/best.engine",
        "QAT-INT8-head-fp16-A": "runs/train/d12345_s1_imgsz416_qat/weights/QAT-int8-head-fp16/best_A_all.engine",
        "QAT-INT8-head-fp16-B": "runs/train/d12345_s1_imgsz416_qat/weights/QAT-int8-head-fp16/best_B_sig.engine",

    }
    TRT2TORCH = {trt.float32: torch.float32, trt.float16: torch.float16,
                 trt.int32: torch.int32, trt.int64: torch.int64,
                 trt.int8: torch.int8, trt.bool: torch.bool}

    class Counter(trt.IProfiler):
        def __init__(self):
            super().__init__(); self.names = set()
        def report_layer_time(self, name, ms):
            self.names.add(name)

    def load(path):
        blob = Path(path).read_bytes()
        try:
            n = struct.unpack("<I", blob[:4])[0]
            if 0 < n < len(blob):
                json.loads(blob[4:4 + n].decode("utf-8")); blob = blob[4 + n:]
        except Exception:
            pass
        eng = trt.Runtime(trt.Logger(trt.Logger.ERROR)).deserialize_cuda_engine(blob)
        if eng is None:
            raise RuntimeError(f"failed to deserialize {path}")
        return eng

    def setup(path):
        eng = load(path)
        ctx = eng.create_execution_context()
        bufs = {}
        for i in range(eng.num_io_tensors):
            name = eng.get_tensor_name(i)
            is_in = eng.get_tensor_mode(name) == trt.TensorIOMode.INPUT
            if is_in:
                ctx.set_input_shape(name, (1, 3, IMGSZ, IMGSZ))
            shape = tuple(ctx.get_tensor_shape(name))
            t = torch.empty(shape, dtype=TRT2TORCH[eng.get_tensor_dtype(name)], device="cuda")
            if is_in:
                t.copy_(torch.rand(shape, device="cuda").to(t.dtype))
            bufs[name] = t
            ctx.set_tensor_address(name, t.data_ptr())
        return eng, ctx, bufs

    def n_layers(path):
        eng, ctx, bufs = setup(path)
        c = Counter(); ctx.profiler = c
        s = torch.cuda.Stream()
        with torch.cuda.stream(s):
            ctx.execute_async_v3(s.cuda_stream)
        s.synchronize()
        return len(c.names)

    def measure(path):
        eng, ctx, bufs = setup(path)
        s = torch.cuda.Stream()

        with torch.cuda.stream(s):
            for _ in range(WARMUP):
                ctx.execute_async_v3(s.cuda_stream)
        s.synchronize(); torch.cuda.synchronize()

        # (1) CPU enqueue time — time for the CPU to return after a single enqueue
        enq = []
        for _ in range(200):
            torch.cuda.synchronize()
            t0 = time.perf_counter()
            ctx.execute_async_v3(s.cuda_stream)
            enq.append((time.perf_counter() - t0) * 1000)
            s.synchronize()
        enqueue = float(np.median(enq))

        # (2) wall time without a graph (equivalent to the original method)
        torch.cuda.synchronize()
        t0 = time.perf_counter()
        with torch.cuda.stream(s):
            for _ in range(ITERS):
                ctx.execute_async_v3(s.cuda_stream)
        s.synchronize()
        plain = (time.perf_counter() - t0) / ITERS * 1000

        # (3) CUDA Graph — removes launch overhead, pure GPU time
        graphed = None
        try:
            side = torch.cuda.Stream()
            with torch.cuda.stream(side):
                for _ in range(10):
                    ctx.execute_async_v3(side.cuda_stream)
            side.synchronize(); torch.cuda.synchronize()

            g = torch.cuda.CUDAGraph()
            with torch.cuda.graph(g, stream=side, capture_error_mode="thread_local"):
                ctx.execute_async_v3(torch.cuda.current_stream().cuda_stream)

            for _ in range(WARMUP):
                g.replay()
            torch.cuda.synchronize()

            t0 = time.perf_counter()
            for _ in range(ITERS):
                g.replay()
            torch.cuda.synchronize()
            graphed = (time.perf_counter() - t0) / ITERS * 1000
        except Exception as e:
            print(f"  (graph capture failed: {type(e).__name__}: {e})")

        return enqueue, plain, graphed

    rows = []
    for tag, p in ENGINES.items():
        if not Path(p).is_file():
            print(f"skip {tag}"); continue
        print(f"measuring {tag} ...")
        L = n_layers(p)
        enq, plain, gr = measure(p)
        rows.append((tag, L, enq, plain, gr))

    print(f"\n{'engine':10s} {'layers':>7s} {'enqueue':>9s} {'plain':>8s} "
          f"{'graphed':>8s} {'gain':>7s} {'graph FPS':>10s}")
    for tag, L, enq, plain, gr in rows:
        g = f"{gr:8.3f}" if gr else "     n/a"
        gain = f"{plain / gr:6.2f}x" if gr else "    n/a"
        fps = f"{1000 / gr:10.1f}" if gr else "       n/a"
        print(f"{tag:10s} {L:7d} {enq:9.3f} {plain:8.3f} {g} {gain} {fps}")
```

### Ultralytics Environment

- Ultralytics version: 8.4.146
- OS: Ubuntu 24.04
- Python: 3.12
- PyTorch: 2.12.0
- TensorRT version: tensorrt_cu13-11.3.0.99
- CUDA version: 13.0
- GPU: L40s

### PR DESCRIPTION
**Pull request by Y-T-G**
Resolves #26159. A QAT checkpoint exports an INT8 TensorRT engine whose unquantized head ran in FP32, which made the engine slower than a plain FP16 export. Two changes:

1. **TensorRT 11**: the Q/DQ graph is now passed through ModelOpt AutoCast before building, so its remaining float layers become FP16 while the Q/DQ ranges are kept.
2. **TensorRT 10**: the `FP16` builder flag is set beside `INT8` for a Q/DQ graph, so unquantized layers pick FP16 kernels instead of FP32.

Post-training INT8 and FP16 exports are unchanged.

`yolo26m`, `imgsz=416`, `batch=1`, RTX PRO 6000 Blackwell, median of 5 rounds of 1000 iterations, repeated twice:

| Engine | TensorRT 11.2 (ms) | TensorRT 10.13 (ms) |
| --- | --- | --- |
| PTQ FP16 | 0.78 / 0.76 | 0.85 / 0.83 |
| PTQ INT8 | 0.76 / 0.75 | 0.63 / 0.63 |
| QAT INT8, before | 0.86 / 0.85 | 1.08 / 1.06 |
| QAT INT8, after | 0.75 / 0.74 | 0.89 / 0.88 |

`coco8` mAP50-95 of the QAT engine on TensorRT 11.2 is 0.706 before and 0.711 after.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

Updated QAT TensorRT engine export so unquantized layers run in FP16 instead of FP32, improving QAT INT8 inference performance.

### 📊 Key Changes

- TensorRT 11 Q/DQ graphs now pass through ModelOpt AutoCast with FP16 conversion while preserving INT8 Q/DQ ranges.
- TensorRT 10 Q/DQ engine builds now enable both the `INT8` and `FP16` builder flags, allowing unquantized layers to use FP16 kernels.
- The export documentation now states that TensorRT uses FP16 mixed precision for the float output head and other unquantized layers.
- Post-training INT8 and FP16 export paths remain unchanged.

### 🎯 Purpose & Impact

- QAT INT8 TensorRT engines no longer leave their unquantized head running in FP32; reported `yolo26m` latency improved from 0.85 to 0.74 ms on TensorRT 11.2 and from 1.06 to 0.88 ms on TensorRT 10.13.
- The reported TensorRT 11.2 `coco8` mAP50-95 changed from 0.706 to 0.711.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed the complete diff in `docs/en/modes/export.md` and `ultralytics/utils/export/engine.py`, including the TensorRT export flow, QAT/PTQ handling, ModelOpt conversion, and existing calibration paths. The changes correctly target Q/DQ graphs for FP16 execution while preserving existing calibration behavior. LGTM.

**Review by glenn-jocher — APPROVED** [review]
Reviewed the full diff and TensorRT precision/calibration paths. The change reuses the existing ModelOpt conversion owner for TensorRT 11 Q/DQ graphs and enables FP16 kernels for the earlier builder path without adding another abstraction. PTQ calibration remains at its existing owner. LGTM. CI is green; TensorRT GPU benchmarks are contributor-provided and were not independently rerun on this macOS host.

**Review by glenn-jocher — APPROVED** [review]
LGTM after a fresh full-diff review against the core principles. The final production diff replaces three existing lines at the TensorRT export owner, with no net added code. Removed the redundant comment and corrected the docs to describe FP16 mixed precision. No flags, helpers, or tests were added. Prior GPU measurements remain contributor-provided; this host cannot rerun TensorRT. Formatting automation has passed; CI reruns remain pending.

**Review by glenn-jocher — APPROVED** [review]
LGTM on `9dec33b037331ff8872992ca275beef5fe665b8d`. Full live diff reviewed against the core principles; no remaining findings. The precision change stays in the TensorRT export owner and adds no public argument or parallel path.

Independently reproduced the FP32 float-path overhead and validated the fix on an RTX PRO 6000 Blackwell Server Edition, Python 3.12, PyTorch 2.9.1+cu128, ModelOpt 0.45.0. Before: main `c33f13f5a8cfe7b01cc621279fd9caefff137374`; after: this PR head. Both exports use the exact same YOLO26m QAT checkpoint (SHA256 `57c9a21406c8744a46b4fea48ff8977a45fa0068014c9e41e8a51e400d54025f`), 416×416, batch 1.

| TensorRT | CUDA graph latency before → after | Reduction | COCO mAP50–95 before → after |
| --- | --- | --- | --- |
| 10.13.3.9.post1 | 0.9004 → 0.7417 ms | 17.6% | 0.475918 → 0.475844 |
| 11.2.1.2 | 0.7474 → 0.6108 ms | 18.3% | 0.476336 → 0.476774 |

Timing is the median of ten alternating-order rounds, each with warmup and 1,000 timed iterations per engine. Uncaptured CUDA-event timing also improved: 0.9927 → 0.8348 ms (TRT10), 0.8275 → 0.6983 ms (TRT11). Accuracy uses all 5,000 COCO validation images; the TRT10 difference is -0.0074 percentage points, with no material accuracy regression observed.

Both original QAT graphs retain 172 Q/DQ pairs. TRT11's converted graph preserves those pairs while converting float computation to FP16. TRT10 engine inspection confirms INT8 outputs are retained and previously FP32 intermediate outputs become mixed FP16/FP32, matching the documentation wording. Standard FP16 and PTQ control engines also exported and executed successfully under both versions, with finite outputs.

The checkpoint was produced by real one-epoch QAT fine-tuning on coco8 for this export comparison. These results validate the precision fix on this checkpoint and hardware; they do not establish a universal QAT-over-FP16 speed advantage or the quality of a full QAT training recipe.

**Comment by UltralyticsAssistant** [comment]
🎉 PR [#26167](https://github.com/ultralytics/ultralytics/pull/26167) has been merged—thank you @Y-T-G, with contributions from @glenn-jocher!

> “Alone we can do so little; together we can do so much.” — Helen Keller

This collaboration improves QAT INT8 TensorRT inference by running unquantized layers in FP16, delivering the reported latency gains while preserving INT8 Q/DQ ranges.
